include/ur_api.h

```diff
@@ -9264,6 +9269,77 @@ urUsmP2PPeerAccessGetInfoExp(
     size_t *pPropSizeRet              ///< [out][optional] pointer to the actual size in bytes of the queried propName.
 );
 
+#if !defined(__GNUC__)
+#pragma endregion
+#endif
+// Intel 'oneAPI' Unified Runtime Experimental API for enqueuing work through native APIs
+#if !defined(__GNUC__)
+#pragma region native enqueue(experimental)
+#endif
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Native enqueue properties
+typedef uint32_t ur_native_enqueue_flags_t;
+typedef enum ur_native_enqueue_flag_t {
+    UR_NATIVE_ENQUEUE_FLAG_TBD = UR_BIT(0), ///< reserved for future use.
+    /// @cond
+    UR_NATIVE_ENQUEUE_FLAG_FORCE_UINT32 = 0x7fffffff
+    /// @endcond
+
+} ur_native_enqueue_flag_t;
+/// @brief Bit Mask for validating ur_native_enqueue_flags_t
+#define UR_NATIVE_ENQUEUE_FLAGS_MASK 0xfffffffe
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Native enqueue properties
+typedef struct ur_native_enqueue_properties_t {
+    ur_structure_type_t stype;       ///< [in] type of this structure, must be
+                                     ///< ::UR_STRUCTURE_TYPE_NATIVE_ENQUEUE_PROPERTIES
+    void *pNext;                     ///< [in,out][optional] pointer to extension-specific structure
+    ur_native_enqueue_flags_t flags; ///< [in] native enqueue flags
+
+} ur_native_enqueue_properties_t;
+
+///////////////////////////////////////////////////////////////////////////////
+/// @brief Function enqueueing work through the native API to be executed
+///        immediately.
+typedef void (*ur_native_enqueue_function_t)(
```

Comment:
There needs to be some way for the adapter to pass data to this callback. For example, L0 usually requires the WaitList events to be part of the "native enqueue":
https://github.com/oneapi-src/unified-runtime/blob/main/source/adapters/level_zero/kernel.cpp#L256

Reply:
That's a good idea. So maybe the callback API should take something like:

```
typedef void (ur_native_enqueue_function_t*)(ur_queue_handle_t, uint32_t numEventsInWaitList,
ur_event_handle_t *inEvents);
```

Do you think we should also have an out event? ie:

```
typedef void (ur_native_enqueue_function_t*)(ur_queue_handle_t, uint32_t numEventsInWaitList,
ur_event_handle_t *inEvents, ur_event_handle_t *outEvents);
```

Reply:
> Do you think we should also have an out event? ie:

Yes, definitely. We also need an L0 command list (the equivalent of a cuda stream), but there's no equivalent UR type.
Igor started a refactoring work on L0 that uses a similar abstraction, see here:
https://github.com/igchor/unified-runtime/blob/queue_v2/source/adapters/level_zero/kernel.cpp#L202

```
  auto L0Enqueue = [&](ze_command_list_handle_t CommandList,
                       ze_event_handle_t SignalEvent, uint32_t WaitListLength,
                       ze_event_handle_t *WaitList) {
    ZE2UR_CALL(zeCommandListAppendLaunchKernel,
               (CommandList, ZeKernel, &ZeThreadGroupDimensions, SignalEvent,
                WaitListLength, WaitList));
    return UR_RESULT_SUCCESS;
  };

  return Queue->V2QueueDispatcher->enqueue(
      OutEvent, NumEventsInWaitList, EventWaitList,
      v2::CommandListPreference::Compute, std::move(L0Enqueue));
```

I'm not sure if there's a clean wait of passing this sort of adapter-specific stuff. Maybe we need to just pass a `void *` here. Alternatively, we could do a union to make this a bit less error-prone:
```
struct ur_enqueue_native_data {
  ur_adapter_backend_t backend;
  union {
     {
        void *CommandList;
     } level_zero;
  };
}
```

I think we need to try and create an implementation to see what's really required :)

Reply:
Yep I will crack on with implementation and keep you in the loop